compiler/lib/src/main/scala/analysis/Semantics/ResolveTopology/MatchedPortNumbering.scala

```diff
@@ -161,29 +187,52 @@ object MatchedPortNumbering {
       val pii = PortInstanceIdentifier(ci, pi)
       val cs = t.getConnectionsAt(pii).toList.sorted
       Result.foldLeft (cs) (empty) ((m, c) => {
-        val piiRemote = c.getOtherEndpoint(pi).port
-        val ciRemote = piiRemote.componentInstance
-        m.get(ciRemote) match {
-          case Some(cPrev) => Left(
-            SemanticError.DuplicateMatchedConnection(
-              c.getLoc,
-              cPrev.getLoc,
-              portMatching.getLoc
+        if(c.isUnmatched)
+          Right(m)
+        else {
+          val piiRemote = c.getOtherEndpoint(pi).port
+          val ciRemote = piiRemote.componentInstance
+          m.get(ciRemote) match {
+            case Some(cPrev) => Left(
+              SemanticError.DuplicateMatchedConnection(
+                c.getLoc,
+                cPrev.getLoc,
+                portMatching.getLoc
+              )
             )
-          )
-          case None => Right(m + (ciRemote -> c))
+            case None => Right(m + (ciRemote -> c))
+          }
         }
       })
     }
+
+    // Computes the set of used ports for all connections at a specific port instance
+    def computeUsedPortNumbers(pi: PortInstance): Result.Result[Set[Int]] = {
+      val empty: Set[Int] = Set()
+      val pii = PortInstanceIdentifier(ci, pi)
+      val cs = t.getConnectionsAt(pii).toList.sorted
+      Result.foldLeft (cs) (empty) ((s, c) => {
+        val used = t.getUsedPortNumbers(pi, List(c))
+        val intersection = s.intersect(used)
+        // Do not allow duplicate port numbers
+        if intersection.nonEmpty
```

Comment:
Is it possible for this intersection to contain more than one element?

Reply:
I think a case where the intersection would contain more than one element is when a connections is using 2 or more port numbers that are already in the accumulator set. In this case I use `nonEmpty` to check if the intersection contains at least one element. If it does, then that means that there is a duplicate (multiple connections are using the same port number) and we throw and error.

Reply:
I don't see how a connection can use 2 port numbers. A connection has the form `a.b[i] -> c.d[j]`, where `i` and `j` are numbers.

Reply:
We are thinking of extending connections to allow ranges, e.g., a.b[0..5] -> c.d[0..5], but we don't allow that yet.

Reply:
I was thinking of the case where i and j are different numbers, so the connection would be using 2 port numbers. Does that make sense?

Reply:
Oh, actually never mind. Only one port number would be used since we are dealing with one port instance at a time.  Sorry for the confusion and to clear this up, the intersection would contain only one element!

Reply:
Update: refactor to do something similar to `checkDuplicateOutputConnections`